source/ftrack_api/uploader.py

```diff
@@ -0,0 +1,150 @@
+# :coding: utf-8
+# :copyright: Copyright (c) 2024 ftrack
+
+import logging
+import math
+import os
+from typing import IO, Awaitable, Callable, TYPE_CHECKING, List, Optional
+import anyio
+
+import anyio.from_thread
+import httpx
+from ._http import ssl_context
+
+if TYPE_CHECKING:
+    from ftrack_api.session import Session
+
+
+SIZE_MEGABYTES = 1024**2
+SIZE_GIGABYTES = 1024**3
+MAX_PARTS = 10000
+
+logger = logging.getLogger(__name__)
+
+
+def get_chunk_size(file_size: int) -> int:
+    chunk_profiles = [
+        (0, 8),
+        (0.1, 16),
+        (1, 32),
+        (8, 64),
+        (64, 128),
+    ]
+
+    for min_size_in_gb, chunk_size_in_mb in reversed(chunk_profiles):
+        if file_size >= min_size_in_gb * SIZE_GIGABYTES:
+            return chunk_size_in_mb * SIZE_MEGABYTES
+
+    raise ValueError("Invalid file size.")
+
+
+async def back_off(func: Callable[..., Awaitable], *args, retries=5, delay=5):
+    for i in range(retries):
+        try:
+            return await func(*args)
+        except Exception as e:
+            if i == retries - 1:
+                raise e
+
+            sleep_time = delay * 2**i
+            logger.warn(f"Retrying in {sleep_time}s")
+            await anyio.sleep(sleep_time)
+
+
+class Uploader:
+    max_concurrency: int = int(os.environ.get("FTRACK_UPLOAD_MAX_CONCURRENCY", 5))
+
+    def __init__(
+        self,
+        session: "Session",
+        component_id: str,
+        file_name: str,
+        file_size: int,
+        file: "IO",
+        checksum: Optional[str],
+    ):
+        self.session = session
+        self.component_id = component_id
+        self.file = file
+        self.file_name = file_name
+        self.file_size = file_size
+        self.checksum = checksum
+
+        self.chunk_size = get_chunk_size(self.file_size)
+        self.parts_count = math.ceil(self.file_size / self.chunk_size)
+
+        if self.parts_count > MAX_PARTS:
+            raise ValueError("File is too big.")
+
+        self.upload_id: Optional[str] = None
```

Comment:
these are only used within `start` perhaps they do not need to be instance variables?